Fix leaking log-level set on root logger in LoggerTests

### CI Link

https://gradle-enterprise.elastic.co/s/ssqi5r5klvhsm

### Repro line

./gradlew ':server:test' --tests "org.elasticsearch.snapshots.SnapshotResiliencyTests*" -Druntime.java=21  -Dtests.es.logger.level=DEBUG

### Does it reproduce?

Yes

### Applicable branches

main

### Failure history

_No response_

### Failure excerpt

```org.elasticsearch.snapshots.SnapshotResiliencyTests > testIllegalArgumentExceptionLogging FAILED
    java.lang.AssertionError: expected to see INFO log but did not
    Expected: <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([782F1BB9840950FE:32EBB1FE53CD0BB8]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.elasticsearch.test.MockLogAppender$SeenEventExpectation.assertMatched(MockLogAppender.java:127)
```

The following 4 tests started failing yesterday in `SnapshotResiliencyTests` due to an incorrectly set log-level.  
```
Tests with failures:
 - org.elasticsearch.snapshots.SnapshotResiliencyTests.testIndexNotFoundExceptionLogging
 - org.elasticsearch.snapshots.SnapshotResiliencyTests.testFullSnapshotUnassignedShards
 - org.elasticsearch.snapshots.SnapshotResiliencyTests.testIllegalArgumentExceptionLogging
 - org.elasticsearch.snapshots.SnapshotResiliencyTests.testSnapshotNameAlreadyInUseExceptionLogging
 ```

 It seems that in this [PR](https://github.com/elastic/elasticsearch/pull/105020) we updated the log-level of the[ root logger](https://github.com/elastic/elasticsearch/blob/main/server/src/test/java/org/elasticsearch/common/logging/LoggersTests.java#L88) as part of the tests, so if it happens that the `SnapshotResiliencyTests` are executed **after** the `LoggersTests`, the log-level for the `SnapshotsService` logger will be set to debug causing the failures above.  

We should probably reset the log-level after the tests, as we do in `InternalEngineTests#testMergeThreadLogging`  for example

--- a/qa/restricted-loggers/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
+++ b/qa/restricted-loggers/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
@@ -69,19 +69,14 @@ public class LoggersTests extends ESTestCase {
         assertThat(Loggers.RESTRICTED_LOGGERS, hasSize(greaterThan(0)));
 
         for (String restricted : Loggers.RESTRICTED_LOGGERS) {
-            Level rootLevel = LogManager.getRootLogger().getLevel();
+            TestLoggers.runWithLoggersRestored(() -> {
+                // 'org.apache.http' is an example of a restricted logger,
+                // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
+                // and the parent logger is `org.apache`.
+                Logger restrictedLogger = LogManager.getLogger(restricted);
+                Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
+                Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
 
-            // 'org.apache.http' is an example of a restricted logger,
-            // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
-            // and the parent logger is `org.apache`.
-            Logger restrictedLogger = LogManager.getLogger(restricted);
-            Level restrictedLoggerLevel = restrictedLogger.getLevel();
-            Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
-            Level restrictedComponentLevel = restrictedComponent.getLevel();
-            Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
-            Level parentLoggerLevel = parentLogger.getLevel();
-
-            try {
                 Loggers.setLevel(restrictedLogger, Level.INFO);
                 assertHasINFO(restrictedLogger, restrictedComponent);
 
@@ -104,13 +99,7 @@ public class LoggersTests extends ESTestCase {
                 Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG);
                 assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
                 assertHasINFO(restrictedComponent, restrictedLogger);
-            } finally {
-                // make sure all levels are restored
-                Loggers.setLevel(LogManager.getRootLogger(), rootLevel);
-                Loggers.setLevel(restrictedLogger, restrictedLoggerLevel);
-                Loggers.setLevel(restrictedComponent, restrictedComponentLevel);
-                Loggers.setLevel(parentLogger, parentLoggerLevel);
-            }
+            });
         }
     }
 

--- a/qa/restricted-loggers/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
+++ b/qa/restricted-loggers/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
@@ -69,36 +69,48 @@ public class LoggersTests extends ESTestCase {
         assertThat(Loggers.RESTRICTED_LOGGERS, hasSize(greaterThan(0)));
 
         for (String restricted : Loggers.RESTRICTED_LOGGERS) {
+            Level rootLevel = LogManager.getRootLogger().getLevel();
 
             // 'org.apache.http' is an example of a restricted logger,
             // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
             // and the parent logger is `org.apache`.
             Logger restrictedLogger = LogManager.getLogger(restricted);
+            Level restrictedLoggerLevel = restrictedLogger.getLevel();
             Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
+            Level restrictedComponentLevel = restrictedComponent.getLevel();
             Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
+            Level parentLoggerLevel = parentLogger.getLevel();
 
-            Loggers.setLevel(restrictedLogger, Level.INFO);
-            assertHasINFO(restrictedLogger, restrictedComponent);
+            try {
+                Loggers.setLevel(restrictedLogger, Level.INFO);
+                assertHasINFO(restrictedLogger, restrictedComponent);
 
-            for (Logger log : List.of(restrictedComponent, restrictedLogger)) {
-                // DEBUG is rejected due to restriction
-                Loggers.setLevel(log, Level.DEBUG);
+                for (Logger log : List.of(restrictedComponent, restrictedLogger)) {
+                    // DEBUG is rejected due to restriction
+                    Loggers.setLevel(log, Level.DEBUG);
+                    assertHasINFO(restrictedComponent, restrictedLogger);
+                }
+
+                // OK for parent `org.apache`, but restriction is enforced for restricted descendants
+                Loggers.setLevel(parentLogger, Level.DEBUG);
+                assertEquals(Level.DEBUG, parentLogger.getLevel());
                 assertHasINFO(restrictedComponent, restrictedLogger);
+
+                // Inheriting DEBUG of parent `org.apache` is rejected
+                Loggers.setLevel(restrictedLogger, (Level) null);
+                assertHasINFO(restrictedComponent, restrictedLogger);
+
+                // DEBUG of root logger isn't propagated to restricted loggers
+                Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG);
+                assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
+                assertHasINFO(restrictedComponent, restrictedLogger);
+            } finally {
+                // make sure all levels are restored
+                Loggers.setLevel(LogManager.getRootLogger(), rootLevel);
+                Loggers.setLevel(restrictedLogger, restrictedLoggerLevel);
+                Loggers.setLevel(restrictedComponent, restrictedComponentLevel);
+                Loggers.setLevel(parentLogger, parentLoggerLevel);
             }
-
-            // OK for parent `org.apache`, but restriction is enforced for restricted descendants
-            Loggers.setLevel(parentLogger, Level.DEBUG);
-            assertEquals(Level.DEBUG, parentLogger.getLevel());
-            assertHasINFO(restrictedComponent, restrictedLogger);
-
-            // Inheriting DEBUG of parent `org.apache` is rejected
-            Loggers.setLevel(restrictedLogger, (Level) null);
-            assertHasINFO(restrictedComponent, restrictedLogger);
-
-            // DEBUG of root logger isn't propagated to restricted loggers
-            Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG);
-            assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
-            assertHasINFO(restrictedComponent, restrictedLogger);
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
+++ b/server/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
@@ -58,36 +58,48 @@ public class LoggersTests extends ESTestCase {
 
     public void testSetLevelWithRestrictions() {
         for (String restricted : restrictedLoggers) {
+            Level rootLevel = LogManager.getRootLogger().getLevel();
 
             // 'org.apache.http' is an example of a restricted logger,
             // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
             // and the parent logger is `org.apache`.
             Logger restrictedLogger = LogManager.getLogger(restricted);
+            Level restrictedLoggerLevel = restrictedLogger.getLevel();
             Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
+            Level restrictedComponentLevel = restrictedComponent.getLevel();
             Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
+            Level parentLoggerLevel = parentLogger.getLevel();
 
-            Loggers.setLevel(restrictedLogger, Level.INFO, restrictedLoggers);
-            assertHasINFO(restrictedLogger, restrictedComponent);
+            try {
+                Loggers.setLevel(restrictedLogger, Level.INFO, restrictedLoggers);
+                assertHasINFO(restrictedLogger, restrictedComponent);
 
-            for (Logger log : List.of(restrictedComponent, restrictedLogger)) {
-                // DEBUG is rejected due to restriction
-                Loggers.setLevel(log, Level.DEBUG, restrictedLoggers);
+                for (Logger log : List.of(restrictedComponent, restrictedLogger)) {
+                    // DEBUG is rejected due to restriction
+                    Loggers.setLevel(log, Level.DEBUG, restrictedLoggers);
+                    assertHasINFO(restrictedComponent, restrictedLogger);
+                }
+
+                // OK for parent `org.apache`, but restriction is enforced for restricted descendants
+                Loggers.setLevel(parentLogger, Level.DEBUG, restrictedLoggers);
+                assertEquals(Level.DEBUG, parentLogger.getLevel());
                 assertHasINFO(restrictedComponent, restrictedLogger);
+
+                // Inheriting DEBUG of parent `org.apache` is rejected
+                Loggers.setLevel(restrictedLogger, null, restrictedLoggers);
+                assertHasINFO(restrictedComponent, restrictedLogger);
+
+                // DEBUG of root logger isn't propagated to restricted loggers
+                Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG, restrictedLoggers);
+                assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
+                assertHasINFO(restrictedComponent, restrictedLogger);
+            } finally {
+                // make sure all levels are restored
+                Loggers.setLevel(LogManager.getRootLogger(), rootLevel);
+                Loggers.setLevel(restrictedLogger, restrictedLoggerLevel);
+                Loggers.setLevel(restrictedComponent, restrictedComponentLevel);
+                Loggers.setLevel(parentLogger, parentLoggerLevel);
             }
-
-            // OK for parent `org.apache`, but restriction is enforced for restricted descendants
-            Loggers.setLevel(parentLogger, Level.DEBUG, restrictedLoggers);
-            assertEquals(Level.DEBUG, parentLogger.getLevel());
-            assertHasINFO(restrictedComponent, restrictedLogger);
-
-            // Inheriting DEBUG of parent `org.apache` is rejected
-            Loggers.setLevel(restrictedLogger, null, restrictedLoggers);
-            assertHasINFO(restrictedComponent, restrictedLogger);
-
-            // DEBUG of root logger isn't propagated to restricted loggers
-            Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG, restrictedLoggers);
-            assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
-            assertHasINFO(restrictedComponent, restrictedLogger);
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
+++ b/server/src/test/java/org/elasticsearch/common/logging/LoggersTests.java
@@ -58,19 +58,14 @@ public class LoggersTests extends ESTestCase {
 
     public void testSetLevelWithRestrictions() {
         for (String restricted : restrictedLoggers) {
-            Level rootLevel = LogManager.getRootLogger().getLevel();
+            TestLoggers.runWithLoggersRestored(() -> {
+                // 'org.apache.http' is an example of a restricted logger,
+                // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
+                // and the parent logger is `org.apache`.
+                Logger restrictedLogger = LogManager.getLogger(restricted);
+                Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
+                Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
 
-            // 'org.apache.http' is an example of a restricted logger,
-            // a restricted component logger would be `org.apache.http.client.HttpClient` for instance,
-            // and the parent logger is `org.apache`.
-            Logger restrictedLogger = LogManager.getLogger(restricted);
-            Level restrictedLoggerLevel = restrictedLogger.getLevel();
-            Logger restrictedComponent = LogManager.getLogger(restricted + ".component");
-            Level restrictedComponentLevel = restrictedComponent.getLevel();
-            Logger parentLogger = LogManager.getLogger(restricted.substring(0, restricted.lastIndexOf('.')));
-            Level parentLoggerLevel = parentLogger.getLevel();
-
-            try {
                 Loggers.setLevel(restrictedLogger, Level.INFO, restrictedLoggers);
                 assertHasINFO(restrictedLogger, restrictedComponent);
 
@@ -93,13 +88,7 @@ public class LoggersTests extends ESTestCase {
                 Loggers.setLevel(LogManager.getRootLogger(), Level.DEBUG, restrictedLoggers);
                 assertEquals(Level.DEBUG, LogManager.getRootLogger().getLevel());
                 assertHasINFO(restrictedComponent, restrictedLogger);
-            } finally {
-                // make sure all levels are restored
-                Loggers.setLevel(LogManager.getRootLogger(), rootLevel);
-                Loggers.setLevel(restrictedLogger, restrictedLoggerLevel);
-                Loggers.setLevel(restrictedComponent, restrictedComponentLevel);
-                Loggers.setLevel(parentLogger, parentLoggerLevel);
-            }
+            });
         }
     }
 

--- a/test/framework/src/main/java/org/elasticsearch/common/logging/TestLoggers.java
+++ b/test/framework/src/main/java/org/elasticsearch/common/logging/TestLoggers.java
@@ -1,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.core.LoggerContext;
+import org.apache.logging.log4j.core.config.Configuration;
+import org.apache.logging.log4j.core.config.LoggerConfig;
+import org.elasticsearch.common.util.Maps;
+
+import java.util.Collection;
+import java.util.NavigableMap;
+import java.util.function.Predicate;
+
+public class TestLoggers {
+
+    // visible for testing
+    static NavigableMap<String, Level> getLogLevels() {
+        Configuration config = LoggerContext.getContext(false).getConfiguration();
+        return getLogLevels(config.getLoggers().values());
+    }
+
+    private static NavigableMap<String, Level> getLogLevels(Collection<LoggerConfig> configs) {
+        return configs.stream().collect(Maps.toUnmodifiableSortedMap(LoggerConfig::getName, LoggerConfig::getLevel));
+    }
+
+    /**
+     * Util to run a task with the original logger context restored afterwards.
+     * This will reset loggers to their previous log levels and remove any additional loggers configured while running task.
+     */
+    public static void runWithLoggersRestored(Runnable task) {
+        Configuration config = LoggerContext.getContext(false).getConfiguration();
+        Collection<LoggerConfig> configs = config.getLoggers().values();
+        var levels = getLogLevels(configs);
+        try {
+            task.run();
+        } finally {
+            // remove any added logger
+            configs.stream().map(LoggerConfig::getName).filter(Predicate.not(levels::containsKey)).forEach(config::removeLogger);
+            // restore original levels (in the right order)
+            levels.forEach((logger, level) -> Loggers.setLevel(LogManager.getLogger(logger), level));
+        }
+    }
+}

--- a/test/framework/src/test/java/org/elasticsearch/common/logging/TestLoggersTests.java
+++ b/test/framework/src/test/java/org/elasticsearch/common/logging/TestLoggersTests.java
@@ -1,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.NavigableMap;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class TestLoggersTests extends ESTestCase {
+
+    public void testRunWithLoggersRestored() {
+        NavigableMap<String, Level> levels = TestLoggers.getLogLevels();
+
+        TestLoggers.runWithLoggersRestored(() -> {
+            Loggers.setLevel(LogManager.getRootLogger(), Level.WARN);
+            Loggers.setLevel(LogManager.getLogger(TestLoggers.class), Level.DEBUG);
+        });
+
+        assertThat(TestLoggers.getLogLevels(), equalTo(levels));
+    }
+}

